### PR TITLE
Portability improvements, first steps at getting neovim on x86_64 SunOS

### PR DESCRIPTION
**Pull request by pekdon**
Small portability fixes to get neovim compiling on SunOS. Solaris < 11 and Illumos still lack a forktpy fallback (work in progress) but these changes should be mostly generic for systems lacking strnlen, pty.h and cfsetspeed.

Ran the test suite on Linux, all tests passing.

### CODE REVIEW

**Comment by bfredl** [comment]
> Small portability fixes to get neovim compiling on SunOS. Solaris < 11

Does this imply than neovim already builds on Solaris 11.4?

**Comment by pekdon** [comment]
> > Small portability fixes to get neovim compiling on SunOS. Solaris < 11
> 
> Does this imply than neovim already builds on Solaris 11.4?

Not when it's taken directly from GIT where the dependencies are taken from the repository, luajit fails. The WIP pkgsrc variant compiles with these patches (and forkpty one not included) on Solaris 10, and should work on Solaris 11 where forkpty exists. The < Solaris11/Illumos comment was referring to forkpty only.

**Comment by clason** [comment]
Just to make this clear early: I don't think we're interested in supporting (and certainly not maintaining) older versions of SunOS. Even then, this may break anytime we update, for example bump a dependency.

**Comment by pekdon** [comment]
> Just to make this clear early: I don't think we're interested in supporting (and certainly not maintaining) older versions of SunOS. Even then, this may break anytime we update, for example to bump a dependency.

Yep, understandable. As there are a bunch of str* compat functions already, thought adding strnlen to the list would be ok?

The rest of the changes seems to be required for Illumos derivates too.

File: src/nvim/strings.c [hunk 1]
```diff
@@ -394,6 +394,18 @@ void del_trailing_spaces(char_u *ptr)
   }
 }
 
+#if !defined(HAVE_STRNLEN)
+size_t vim_strnlen(const char *s, size_t n)
```
Comment (bfredl) [inline src/nvim/strings.c:398]: if we add this, by our conventions it should be `xstrnlen`

Comment (pekdon) [reply src/nvim/strings.c]: Renamed to xstrnlen

**Comment by pekdon** [comment]
Looking at the test runs I notice:

```
src/nvim/os/pty_process_unix.c:17:  Found C system header after other header. Should be: c system, c++ system, other.  [build/include_order] [4]
```

users.c and eval.c does the same, assuming errors.json (or similar) would need an exception.

**Comment by clason** [comment]
Lint checks (near) added lines. That not all legacy code conforms with the Neovim code style is no excuse to add _new_ lines that don't ;)

**Comment by pekdon** [comment]
> Lint checks (near) added lines. That not all legacy code conforms with the Neovim code style is no excuse to add _new_ lines that don't ;)

Sure, how do you suggest I get access to the defines for checking if a header is present without including the header that defines it?

I guess the build could be changed to use -DHAVE_XXX_H instead of having that in a header file, is that something you would want to see?

**Comment by clason** [comment]
It complains about the _ordering_.

**Comment by pekdon** [comment]
> It complains about the _ordering_.

Yes, but I can't include <pty.h> if it doesn't exist and config.h contains the info about that.

I'll just skip the check for pty.h and not include it if __sun is defined instead.

**Comment by dundargoc** [comment]
Is it possible to move `#include "auto/config.h"` below the `#if` block? So that it will look like the following:

```c
// forkpty is not in POSIX, so headers are platform-specific
#if defined(__FreeBSD__) || defined(__DragonFly__)
# include <libutil.h>
#elif defined(__OpenBSD__) || defined(__NetBSD__) || defined(__APPLE__)
# include <util.h>
#elif defined(HAVE_PTY_H)
# include <pty.h>
#endif

#include "auto/config.h"
```

This will make the warning go away.

**Comment by pekdon** [comment]
Not without adding -DHAVE_PTY_H to the compiler or similar @dundargoc, as it would make HAVE_PTY_H undefined at the time of the check and cause pty.h to not be included.

I just changed the code to skip including pty.h if __sun is defined, doesn't exist on any of the SunOS machines I have access to.

**Comment by pekdon** [comment]
As opinions seems to differ here I'll leave the pty.h check out and just use __sun if/until somebody clearly states it should go back in @jamessan.

**Comment by jamessan** [comment]
Yeah, I'm fine with the change you made, @pekdon, but we _do_ sometimes decide to accept new lint issues if there's a decent reason behind it.

**Comment by pekdon** [comment]
The Ubuntu lint is collateral due to edits close to the STRCPY macro:
```
src/nvim/vim.h:223:  Use xstrlcpy or snprintf instead of strcpy  [runtime/printf] [4]
```
Not sure on how to interpret the failed mac build however.

**Comment by dundargoc** [comment]
> The Ubuntu lint is collateral due to edits close to the STRCPY macro:
> 
> ```
> src/nvim/vim.h:223:  Use xstrlcpy or snprintf instead of strcpy  [runtime/printf] [4]
> ```
> 
> Not sure on how to interpret the failed mac build however.

Likely a flaky test. We can try rerunning it and see if it works next time.

**Comment by pekdon** [comment]
@dundargoc passed this time around, windows test did however fail instead.

Let me know if there is something that I should look at/improve/change.

**Pull request merged**